### PR TITLE
Improve documentation for `TaskAPI`

### LINKED ISSUE #8694: Add documentation for Task API
Add API documentation for Task(contribution) & related components and potentially an extension guide.

### PR DESCRIPTION
**Pull request by tortmayr**
Fixes #8694
Signed-off-by: Tobias Ortmayr <tortmayr@eclipsesource.com>
Contributed on behalf of STMicroelectronics

#### What it does

Adds/refines documentation for components of the Task API.
#### How to test

Documentation changes only
#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/task/src/browser/task-contribution.ts [hunk 1]
```diff
@@ -21,26 +21,79 @@ import { WaitUntilEvent, Emitter } from '@theia/core/lib/common/event';
...
+ * The {@link TaskResolverRegistry} is the common component for registration and provision of
+ * {@link TaskResolver}s. Theia will collect all {@link TaskContribution}s and invoke {@link TaskContribution#registerResolvers}
+ * for each contribution.
+ */
 @injectable()
 export class TaskResolverRegistry {
 
     protected readonly onWillProvideTaskResolverEmitter = new Emitter<WaitUntilEvent>();
+    /**
+     * Emit when the registry provides a registed resolver. i.e. when the {@link TaskResolverRegistry#getResolver}
```
Comment (vince-fugnitto) [inline packages/task/src/browser/task-contribution.ts:94]:
```suggestion
     * Emit when the registry provides a registered resolver. i.e. when the {@link TaskResolverRegistry#getResolver}
```

File: packages/task/src/node/task-manager.ts [hunk 2]
```diff
@@ -20,6 +20,9 @@ import { Task } from './task';
 
 // inspired by process-manager.ts
 
+/**
+ * The {@link TaskManager} is the common component responsible for mananging running tasks.
```
Comment (vince-fugnitto) [inline packages/task/src/node/task-manager.ts:24]:
```suggestion
 * The {@link TaskManager} is the common component responsible for managing running tasks.
```

File: packages/task/src/node/task-manager.ts [hunk 3]
```diff
@@ -29,12 +32,23 @@ export class TaskManager implements BackendApplicationContribution {
     protected readonly tasksPerCtx: Map<string, Task[]> = new Map();
     /** each task has this unique task id, for this back-end */
     protected id: number = -1;
+    /**
+     * Emit when a registed task is deleted.
```
Comment (vince-fugnitto) [inline packages/task/src/node/task-manager.ts:36]:
```suggestion
     * Emit when a registered task is deleted.
```

File: packages/task/src/node/task-manager.ts [hunk 4]
```diff
@@ -51,13 +65,22 @@ export class TaskManager implements BackendApplicationContribution {
...
+     */
     get(id: number): Task | undefined {
         return this.tasks.get(id);
     }
 
     /**
      * Returns all running tasks. If a context is provided, filter-down to
-     * only tasks started from that context
+     * only tasks started from that context.
+     * @param ctx the task extecution context.
```
Comment (vince-fugnitto) [inline packages/task/src/node/task-manager.ts:81]:
```suggestion
     * @param ctx the task execution context.
```

File: packages/task/src/node/task-manager.ts [hunk 5]
```diff
@@ -51,13 +65,22 @@ export class TaskManager implements BackendApplicationContribution {
...
+     *
+     * @returns all running tasks for the given context or `undefined` if no tasks are registed for the given context.
```
Comment (vince-fugnitto) [inline packages/task/src/node/task-manager.ts:83]:
```suggestion
     * @returns all running tasks for the given context or `undefined` if no tasks are registered for the given context.
```

File: packages/task/src/node/task-runner.ts [hunk 6]
```diff
@@ -13,27 +13,44 @@
...
+     * @param taskConfig the task configuration that should be executed
+     * @param ctx the execution context
+     *
+     * @returns a promise of the (currently running) {@link Task}.
+     */
+    run(tskConfig: TaskConfiguration, ctx?: string): Promise<Task>;
 }
 
+/**
+ * The {@link TaskRunnerRegistry} is the common componet for the registration and provisioning of
```
Comment (vince-fugnitto) [inline packages/task/src/node/task-runner.ts:49]:
```suggestion
 * The {@link TaskRunnerRegistry} is the common component for the registration and provisioning of
```

File: packages/task/src/node/task-runner.ts [hunk 7]
```diff
@@ -13,27 +13,44 @@
...
+ * {@link TaskRunner}s. Theia will collect all {@link TaskRunner}s and invoke {@link TaskRunnerContribution#registerRunner}
+ * for each contributon. The `TaskServer` will use the runnners provided by this registry to execute `TaskConfiguration`s that
```
Comment (vince-fugnitto) [inline packages/task/src/node/task-runner.ts:51]:
```suggestion
 * for each contribution. The `TaskServer` will use the runnners provided by this registry to execute `TaskConfiguration`s that
```

File: packages/task/src/node/task-runner.ts [hunk 8]
```diff
@@ -13,27 +13,44 @@
...
  * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
  ********************************************************************************/
-
-import { injectable, inject, postConstruct } from 'inversify';
 import { Disposable } from '@theia/core/lib/common/disposable';
+import { inject, injectable, postConstruct } from 'inversify';
+import { TaskConfiguration } from '../common/task-protocol';
 import { ProcessTaskRunner } from './process/process-task-runner';
 import { Task } from './task';
-import { TaskConfiguration } from '../common/task-protocol';
```
Comment (vince-fugnitto) [inline packages/task/src/node/task-runner.ts:21]: I believe these changes are unnecessary.

File: packages/task/src/node/task-runner.ts [hunk 9]
```diff
@@ -13,27 +13,44 @@
...
+/**
+ * A {@link TaskRunner} knows how to run a task configuration of a particular type.
+ */
 export interface TaskRunner {
-    /** Runs a task based on the given task configuration. */
-    run(taskConfig: TaskConfiguration, ctx?: string): Promise<Task>;
+    /**
+     * Runs a task based on the given `TaskConfiguration`.
+     * @param taskConfig the task configuration that should be executed
+     * @param ctx the execution context
```
Comment (vince-fugnitto) [inline packages/task/src/node/task-runner.ts:41]:
```suggestion
     * @param taskConfig the task configuration that should be executed.
     * @param ctx the execution context.
```

**Review by vince-fugnitto — COMMENTED** [review]
I verified the content and it looks good :+1: I have some comments however:

File: packages/task/src/node/task-runner.ts (hunk 8 above)
Comment (tortmayr) [reply packages/task/src/node/task-runner.ts]: Agreed. I removed them

**Comment by tortmayr** [comment]
Thanks for the fast review. I applied your feedback suggestions.

File: packages/task/src/node/task-runner.ts [hunk 10]
```diff
@@ -22,18 +22,36 @@ import { TaskConfiguration } from '../common/task-protocol';
...
+     *
+     * @returns a promise of the (currently running) {@link Task}.
+     */
+    run(tskConfig: TaskConfiguration, ctx?: string): Promise<Task>;
 }
 
+/**
+ * The {@link TaskRunnerRegistry} is the common component for the registration and provisioning of
+ * {@link TaskRunner}s. Theia will collect all {@link TaskRunner}s and invoke {@link TaskRunnerContribution#registerRunner}
+ * for each contribution. The `TaskServer` will use the runnners provided by this registry to execute `TaskConfiguration`s that
```
Comment (vince-fugnitto) [inline packages/task/src/node/task-runner.ts:52]:
```suggestion
 * for each contribution. The `TaskServer` will use the runners provided by this registry to execute `TaskConfiguration`s that
```

File: packages/task/src/browser/task-contribution.ts [hunk 11]
```diff
@@ -77,7 +153,13 @@ export class TaskProviderRegistry {
         this.providers = new Map();
     }
 
-    /** Registers the given Task Provider to return Task Configurations of the specified type. */
+    /**
+     * Registers the given {@link TaskProvider} for task configurations of the specified type
+     * @param type the task configuration type for which the given provider should be registered.
+     * @param provider the `TaskProvider` that should be registered
```
Comment (vince-fugnitto) [inline packages/task/src/browser/task-contribution.ts:159]:
```suggestion
     * @param provider the `TaskProvider` that should be registered.
```

File: packages/task/src/browser/task-contribution.ts [hunk 12]
```diff
@@ -77,7 +153,13 @@ export class TaskProviderRegistry {
...
+     *
+     * @returns a `Disposable` that can be invoked to unregister the given resolver
```
Comment (vince-fugnitto) [inline packages/task/src/browser/task-contribution.ts:161]:
```suggestion
     * @returns a `Disposable` that can be invoked to unregister the given resolver.
```

File: packages/task/src/browser/task-contribution.ts [hunk 13]
```diff
@@ -86,12 +168,24 @@ export class TaskProviderRegistry {
         };
     }
 
+    /**
+     * Retrieves the {@link TaskProvider} registered for the given type task configuration type.
+     * If there is already a `TaskProvider` registered for the specified type the registration will
+     * be overwritten with the new value.
+     * @param type the task configuration type
```
Comment (vince-fugnitto) [inline packages/task/src/browser/task-contribution.ts:175]:
```suggestion
     * @param type the task configuration type.
```

File: packages/task/src/node/task-manager.ts [hunk 14]
```diff
@@ -29,12 +32,23 @@ export class TaskManager implements BackendApplicationContribution {
...
         @inject(ILogger) @named('task') protected readonly logger: ILogger
     ) { }
 
+    /**
+     * Registers a new task (in the given context if present). Each registered
+     * task is considered to be currently running.
+     * @param task the new task
+     * @param ctx  the provided context
+     *
+     * @returns the registration id for the given task
```
Comment (vince-fugnitto) [inline packages/task/src/node/task-manager.ts:50]:
```suggestion
     * @param task the new task.
     * @param ctx  the provided context.
     *
     * @returns the registration id for the given task.
```

File: packages/task/src/node/task-manager.ts [hunk 15]
```diff
@@ -69,7 +92,10 @@ export class TaskManager implements BackendApplicationContribution {
         }
     }
 
-    /** Deletes a task from the task manager */
+    /**
+     * Delete the given task from the task manager.
+     * @param task the task to delete
```
Comment (vince-fugnitto) [inline packages/task/src/node/task-manager.ts:97]:
```suggestion
     * @param task the task to delete.
```

**Review by vince-fugnitto — APPROVED** [review]
@tortmayr the changes look good, the only thing to address are the typos and missing punctuation.

**Pull request merged**